File: crates/wesc-php/stubs/wesc.php
<?php

// Stubs for the `wesc` PHP extension (crates/wesc-php).
//
// These declarations exist purely so editors and static analysers understand
// the functions the native extension registers. They are never executed — the
// real implementations live in the compiled extension. Keep them in sync with
// `crates/wesc-php/src/lib.rs`.

/**
 * Build the entry points and return the HTML output plus the bundled assets.
 *
 * The result is an associative array with three keys:
 * - `html` (`string`): a binary-safe string with the exact output bytes, so you
 *   can `echo` it straight to the response.
 * - `css` (`?string`): the bundled CSS, or `null` when no CSS was produced.
 * - `js` (`?string`): the bundled JS, or `null` when no JS was produced.
 *
 * `$outcss`/`$outjs` still control file output: `null` skips the bundle, a
 * non-empty path bundles and writes the file (and returns it), and an empty
 * string bundles and returns the asset without writing it to disk.
 *
 * @param string[]    $input  Entry point file paths. The first entry is the host document.
 * @param string|null $outcss Optional path to write the bundled CSS file (`''` = in-memory only).
 * @param string|null $outjs  Optional path to write the bundled JS file (`''` = in-memory only).
 * @param bool        $minify Minify generated JS/CSS assets where supported.
 *
 * @return array{html: string, css: ?string, js: ?string} The HTML output and bundled assets.
 */
function wesc_build(
    array $input,
    ?string $outcss = null,
    ?string $outjs = null,
    bool $minify = false,
): array {}

/**
 * Stream the build to a callable, chunk by chunk, for low-memory output.
 *
 * The callable is invoked with each string chunk as it is produced, then once
 * with `null` to signal the end of the stream. If it throws, the exception
 * propagates out and the build stops.
 *
 * Only the HTML is streamed; the bundled CSS/JS are not returned here. Pass
 * `$outcss`/`$outjs` to write the bundles to disk, or use `wesc_build()` if you
 * need the assets back in memory.
 *
 * @param string[]                   $input    Entry point file paths. The first entry is the host document.
 * @param callable(?string): mixed   $callback Called with each chunk, then `null` at end-of-stream.
 * @param string|null                $outcss   Optional path to write the bundled CSS file.
 * @param string|null                $outjs    Optional path to write the bundled JS file.
 * @param bool                       $minify   Minify generated JS/CSS assets where supported.
 *
 * @return void
 */
function wesc_build_stream(
    array $input,
    callable $callback,
    ?string $outcss = null,
    ?string $outjs = null,
    bool $minify = false,
): void {}

File: crates/wesc-rb/lib/wesc.rb
# frozen_string_literal: true

# WeSC — We are the Superlative Components.
#
# A streaming HTML / web-component bundler. This gem exposes the Rust core (via
# a native extension) so it can run in-process on a Ruby server — no subprocess,
# no WASM.
#
#   require "wesc"
#
#   # One-shot: returns a Wesc::Result with the HTML plus the bundled CSS/JS.
#   result = Wesc.build(["./index.html"], minify: true)
#   result.html # => the rendered HTML (binary String)
#
#   # Streaming: low memory, chunk by chunk. The block gets each chunk as a
#   # String, then `nil` once to signal end-of-stream.
#   Wesc.build_stream(["./index.html"]) { |chunk| ... }
#
# See https://github.com/luwes/wesc for the full documentation.

require_relative "wesc/version"
# The native extension (defines Wesc::Native). Built from crates/wesc-rb via
# rb-sys; see the gem's Rakefile (`rake compile`).
require_relative "wesc/wesc_rb"

module Wesc
  # The result of a one-shot {Wesc.build}: the rendered +html+ plus the bundled
  # +css+/+js+. The CSS/JS are populated when requested via `outcss`/`outjs`
  # (an empty string keeps them in memory only; a path also writes the file) and
  # are `nil` otherwise.
  Result = Struct.new(:html, :css, :js)

  module_function

  # Build the entry points and return a {Wesc::Result}.
  #
  #   result = Wesc.build(["./index.html"], minify: true)
  #   result.html # => the rendered HTML (binary String)
  #   result.css  # => the bundled CSS, or nil
  #   result.js   # => the bundled JS, or nil
  #
  # @param input [Array<String>] entry point paths; the first is the host
  #   document.
  # @param outcss [String, nil] request the bundled CSS. A non-empty path also
  #   writes the file; an empty string keeps it in memory only; `nil` skips it.
  # @param outjs [String, nil] request the bundled JS. A non-empty path also
  #   writes the file; an empty string keeps it in memory only; `nil` skips it.
  # @param minify [Boolean] minify generated JS/CSS assets. Defaults to false.
  # @return [Wesc::Result] the rendered HTML (ASCII-8BIT / binary encoding) plus
  #   the bundled CSS/JS (each `nil` when not requested).
  def build(input, outcss: nil, outjs: nil, minify: false)
    html, css, js = Native.build(Array(input), outcss, outjs, minify ? true : false)
    Result.new(html, css, js)
  end

  # Stream the build to a block, chunk by chunk, for low-memory output.
  #
  # The block is called with each chunk as a String, then once with `nil` to
  # signal the end of the stream. Raising from the block stops the build and the
  # exception propagates out of this method.
  #
  #   Wesc.build_stream(["./index.html"]) do |chunk|
  #     io.write(chunk) unless chunk.nil?
  #   end
  #
  # @param input [Array<String>] entry point paths; the first is the host
  #   document.
  # @param outcss [String, nil] optional path to write the bundled CSS file.
  # @param outjs [String, nil] optional path to write the bundled JS file.
  # @param minify [Boolean] minify generated JS/CSS assets. Defaults to false.
  # @yieldparam chunk [String, nil] each output chunk, then `nil` at end-of-stream.
  # @return [void]
  def build_stream(input, outcss: nil, outjs: nil, minify: false, &block)
    raise ArgumentError, "Wesc.build_stream requires a block" unless block

    Native.build_stream(Array(input), outcss, outjs, minify ? true : false, &block)
  end
end
